### PR TITLE
Cumulative Stalk per BDV 

### PR DESCRIPTION
**Pull request by Brean0**
# Problem

Currently there are 2 problems in Beanstalk related to Deposits

1. Farmers with older Deposits may forfeit Stalk when converting from a Whitelisted Silo token with a higher `seedsPerBdv` to a Whitelisted Silo token with a lower `seedsPerBdv`
2.  `seedsPerBdv` must be static over time for each Whitelisted Silo token
3. `seedsPerBdv` cannot have Decimals

https://beanstalk-farms.notion.site/RFC-Cumulative-Stalk-per-BDV-2ed64ebd9aa6448f92e015c7a254b5c6

### CODE REVIEW

File: protocol/contracts/beanstalk/AppStorage.sol [hunk 1]
```diff
@@ -112,13 +118,15 @@ contract Account {
...
         uint256 wrappedBeans; // DEPRECATED – Replant generalized Internal Balances. Wrapped Beans are now stored at the AppStorage level.
-        mapping(address => mapping(uint32 => Deposit)) deposits; // A Farmer's Silo Deposits stored as a map from Token address to Season of Deposit to Deposit.
+        mapping(address => mapping(uint32 => Deposit)) legacyDeposits; // Legacy Silo Deposits stored as a map from Token address to Season of Deposit to Deposit.
         mapping(address => mapping(uint32 => uint256)) withdrawals; // DEPRECATED - Zero withdraw eliminates a need for withdraw mapping
         SeasonOfPlenty sop; // A Farmer's Season Of Plenty storage.
         mapping(address => mapping(address => uint256)) depositAllowances; // Spender => Silo Token
         mapping(address => mapping(IERC20 => uint256)) tokenAllowances; // Token allowances
         uint256 depositPermitNonces; // A Farmer's current deposit permit nonce
         uint256 tokenPermitNonces; // A Farmer's current token permit nonce
+        mapping(address => mapping(int128 => Deposit)) deposits; //SiloV3 deposits stored as a map from Cumulative Grown Stalk
```
Comment (Brean0) [inline protocol/contracts/beanstalk/AppStorage.sol:128]: could we change the comment to: SiloV3 Deposits stored as a map from Token address to CulmativeGrownStalk to Deposit ?

File: protocol/contracts/beanstalk/AppStorage.sol [hunk 2]
```diff
@@ -184,7 +192,7 @@ contract Storage {
     // Silo stores global level Silo balances.
     struct Silo {
         uint256 stalk; // The total amount of active Stalk (including Earned Stalk, excluding Grown Stalk).
-        uint256 seeds; // The total amount of active Seeds (excluding Earned Seeds).
+        uint256 seeds; // The total amount of active Seeds (excluding Earned Seeds). No longer used.
```
Comment (Brean0) [inline protocol/contracts/beanstalk/AppStorage.sol:195]: can we add a DEPRECATED here

File: protocol/contracts/beanstalk/AppStorage.sol [hunk 3]
```diff
@@ -267,13 +276,27 @@ contract Storage {
          */
         bytes4 selector;
         /*
-         * @dev The Seeds Per BDV that the Silo mints in exchange for Depositing this Token.
+         * @dev The Stalk Per BDV Per Season represents how much Stalk one BDV of the underlaying deposited token grows each season. In the past, this was represented by seeds. This is stored as 1e6, plus stalk is stored as 1e10, so 1 legacy
```
Comment (Brean0) [inline protocol/contracts/beanstalk/AppStorage.sol:279]: underlying

Comment (Brean0) [inline protocol/contracts/beanstalk/AppStorage.sol:279]: underlying

File: protocol/contracts/beanstalk/AppStorage.sol (hunk 2 above)
Comment (silochad) [reply protocol/contracts/beanstalk/AppStorage.sol:195]: Please use `deprecated_seeds`

**Comment by Brean0** [comment]
closed in favor of https://github.com/BeanstalkFarms/Beanstalk/pull/352

File: protocol/contracts/libraries/Silo/LibTokenSilo.sol [hunk 4]
```diff
@@ -121,25 +134,40 @@ library LibTokenSilo {
...
-
-        s.a[account].deposits[token][season].amount += uint128(amount);
-        s.a[account].deposits[token][season].bdv += uint128(bdv);
-
-        emit AddDeposit(account, token, season, amount, bdv);
+        console.log('addDepositToAccount token: ', token);
+        console.log('addDepositToAccount logging grown stalk per bdv:');
+        console.logInt(grownStalkPerBdv);
+        uint256 updatedAmount = s.a[account].deposits[token][grownStalkPerBdv].amount.add(amount.toUint128());
+        s.a[account].deposits[token][grownStalkPerBdv].amount = uint128(updatedAmount);
```
[the remark this answers is not in the document]
Comment (Brean0) [reply protocol/contracts/libraries/Silo/LibTokenSilo.sol:146]: done on brean branch

File: protocol/contracts/libraries/Silo/LibTokenSilo.sol [hunk 5]
```diff
@@ -121,25 +134,40 @@ library LibTokenSilo {
...
+        console.log('addDepositToAccount updatedAmount: ', updatedAmount);
+        uint256 updatedTotalTokenBdv = s.a[account].deposits[token][grownStalkPerBdv].bdv.add(bdv.toUint128());
+        s.a[account].deposits[token][grownStalkPerBdv].bdv = uint128(updatedTotalTokenBdv);
+        console.log('addDepositToAccount bdv: ', bdv);
+
+        console.log('--- s.a[account].deposits[token][grownStalkPerBdv].bdv: ', s.a[account].deposits[token][grownStalkPerBdv].bdv);
+        console.log('--- s.a[account].deposits[token][grownStalkPerBdv].amount: ', s.a[account].deposits[token][grownStalkPerBdv].amount);
+
+        //setup or update the MowStatus for this deposit. We should have _just_ mowed before calling this function.
+        uint256 updatedMowStatusBdv = s.a[account].mowStatuses[token].bdv.add(bdv.toUint128());
```
[the remark this answers is not in the document]
Comment (Brean0) [reply protocol/contracts/libraries/Silo/LibTokenSilo.sol:156]: done on brean branch

File: protocol/contracts/libraries/Convert/LibLambdaConvert.sol [hunk 6]
```diff
@@ -25,6 +25,7 @@ library LibLambdaConvert {
         )
     {
         (inAmount, tokenIn) = convertData.lambdaConvert();
+        LibInternal.mow(msg.sender, tokenIn); //this convert function used to be pure, had to remove pure in order to mow
```
[the remark this answers is not in the document]
Comment (Brean0) [reply protocol/contracts/libraries/Convert/LibLambdaConvert.sol:28]: cannot

File: protocol/contracts/beanstalk/silo/ConvertFacet.sol [hunk 7]
```diff
@@ -62,100 +67,142 @@ contract ConvertFacet is ReentrancyGuard {
...
+        console.logInt(_cumulativeGrownStalk);
+
+
+        //if we attempt to deposit at a half-season (a grown stalk index that would fall between seasons)
+        //then in affect we lose that partial season's worth of stalk when we deposit
+        //so here we need to update grownStalk to be the amount you'd have with the above deposit
+        grownStalk = uint256(LibTokenSilo.calculateStalkFromGrownStalkIndexAndBdv(IERC20(token), _cumulativeGrownStalk, bdv));
+        console.log('_depositTokens grownStalk: ', grownStalk);
+
+        uint256 stalk = bdv.mul(LibTokenSilo.stalkPerBdv(token)).add(grownStalk);
```
[the remark this answers is not in the document]
Comment (Brean0) [reply protocol/contracts/beanstalk/silo/ConvertFacet.sol:200]: done

File: protocol/contracts/beanstalk/init/InitBipNewSilo.sol [hunk 8]
```diff
@@ -0,0 +1,69 @@
...
+        uint32 stalkPerBdvPerSeason,
+        uint32 season
+    );
+    
+    
+    function init() external {
+        console.log("InitBipNewSilo.init() got called");
+        
+        //update all silo info for current Silo-able assets
+
```
[the remark this answers is not in the document]
Comment (Brean0) [reply protocol/contracts/beanstalk/init/InitBipNewSilo.sol:33]: believe this is done

File: protocol/contracts/beanstalk/AppStorage.sol [hunk 9]
```diff
@@ -267,13 +276,27 @@ contract Storage {
...
+        uint32 stalkPerBdv;
+        /*
+         * @dev The last season in which the Cumulative Grown Stalk Per BDV was updated
+         */
+		uint32 lastUpdateSeason;
+        /*
+         * @dev The cumulative amount of grown stalk per BDV for this Silo depositable token.
          */
-        uint32 seeds;
+		int128 lastCumulativeGrownStalkPerBdv;
```
[the remark this answers is not in the document]
Comment (Brean0) [reply protocol/contracts/beanstalk/AppStorage.sol:295]: done

File: protocol/contracts/beanstalk/silo/SiloFacet/Silo.sol [hunk 10]
```diff
@@ -137,31 +117,105 @@ contract Silo is SiloExit {
...
 
-    function __mow(address account) private {
-        // If this `account` has no Seeds, skip to save gas.
-        if (s.a[account].s.seeds == 0) return;
-        LibSilo.mintStalk(account, balanceOfGrownStalk(account));
+    function __mow(address account, address token) private {
+        console.log('__mow, current season:', s.season.current);
+
+        //require that user account seeds be zero
+        require(s.a[account].s.seeds == 0, 'silo migration needed');
```
[the remark this answers is not in the document]
Comment (Brean0) [reply protocol/contracts/beanstalk/silo/SiloFacet/Silo.sol:131]: done, but `isMigrated` cannot be achieved without sacrificing space in mowStatus. 

current solution: check `s.a[account].mowStatuses[token].lastCumulativeGrownStalkPerBdv == _cumulativeGrownStalkPerBdv`

File: protocol/contracts/beanstalk/silo/SiloFacet/Silo.sol [hunk 11]
```diff
@@ -137,31 +117,105 @@ contract Silo is SiloExit {
...
+
+        // If this `account` has no BDV, skip to save gas. Still need to update lastCumulativeGrownStalkPerBdv (happen on initial deposit, since mow is called before any deposit)
+        if (s.a[account].mowStatuses[token].bdv == 0) {
+            console.log('mow status bdv was zero for token: ', token);
+            s.a[account].mowStatuses[token].lastCumulativeGrownStalkPerBdv = LibTokenSilo.cumulativeGrownStalkPerBdv(IERC20(token));
```
[the remark this answers is not in the document]
Comment (Brean0) [reply protocol/contracts/beanstalk/silo/SiloFacet/Silo.sol:136]: cannot, as, we use last culmativeGrownStalkperBdv in balance of grownStalk

**Comment by Brean0** [comment]
in `balanceOfGrownStalk`, we call `libSilo.stalkReward()`, which calls 
- s.a[account].mowStatuses[token].lastCumulativeGrownStalkPerBdv
- LibTokenSilo.cumulativeGrownStalkPerBdv(IERC20(token)), //get latest grown stalk per bdv for this token
- s.a[account].mowStatuses[token].bdv

given that in this is often called where this is already read, think we can save gas by having an internal function to pass in the values

**Pull request closed without merging**